File: src/fs/home-browser.ts
import { homedir } from 'node:os';
import path from 'node:path';
import { realpath, readdir } from 'node:fs/promises';

export type FolderEntry = {
  name: string;
  path: string;
  kind: 'directory' | 'file';
};

export type FolderListing = {
  root: string;
  path: string;
  parent: string | null;
  entries: FolderEntry[];
};

let configuredRoot = path.resolve(homedir());
let configuredRootReal = path.resolve(homedir());

async function resolveAndAssertWithinRoot(targetPath: string): Promise<string> {
  const resolved = path.resolve(targetPath);
  const resolvedReal = await realpath(resolved);
  if (resolvedReal === configuredRootReal || resolvedReal.startsWith(`${configuredRootReal}${path.sep}`)) {
    return resolvedReal;
  }
  throw new Error('Path must be inside the configured base path.');
}

export async function setBrowserRoot(basePath?: string): Promise<string> {
  const resolved = path.resolve(basePath ?? homedir());
  const resolvedReal = await realpath(resolved);
  configuredRoot = resolved;
  configuredRootReal = resolvedReal;
  return configuredRootReal;
}

export function getHomeRoot(): string {
  return configuredRootReal;
}

export async function listFolder(inputPath?: string): Promise<FolderListing> {
  const currentPath = await resolveAndAssertWithinRoot(inputPath ?? configuredRoot);
  const dirents = await readdir(currentPath, { withFileTypes: true });

  const entries = dirents
    .filter((entry) => !entry.name.startsWith('.'))
    .map((entry) => {
      const entryPath = path.join(currentPath, entry.name);
      return {
        name: entry.name,
        path: entryPath,
        kind: entry.isDirectory() ? 'directory' : 'file',
      } as FolderEntry;
    })
    .sort((a, b) => {
      if (a.kind !== b.kind) {
        return a.kind === 'directory' ? -1 : 1;
      }
      return a.name.localeCompare(b.name);
    });

  const parent = currentPath === configuredRootReal ? null : path.dirname(currentPath);

  return {
    root: configuredRootReal,
    path: currentPath,
    parent,
    entries,
  };
}

File: internal/fs/home_browser.go
package fs

import (
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
)

type FolderEntry struct {
	Name string `json:"name"`
	Path string `json:"path"`
	Kind string `json:"kind"`
}

type FolderListing struct {
	Root    string        `json:"root"`
	Path    string        `json:"path"`
	Parent  *string       `json:"parent"`
	Entries []FolderEntry `json:"entries"`
}

var (
	rootMu         sync.RWMutex
	configuredRoot string
	configuredReal string
)

func init() {
	home, err := os.UserHomeDir()
	if err != nil || home == "" {
		home = "/"
	}
	resolved := filepath.Clean(home)
	real := resolved
	if resolvedReal, err := filepath.EvalSymlinks(resolved); err == nil {
		real = resolvedReal
	}
	configuredRoot = resolved
	configuredReal = real
}

func SetBrowserRoot(basePath string) (string, error) {
	if strings.TrimSpace(basePath) == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", err
		}
		basePath = home
	}
	resolved := filepath.Clean(basePath)
	real, err := filepath.EvalSymlinks(resolved)
	if err != nil {
		return "", err
	}

	rootMu.Lock()
	configuredRoot = resolved
	configuredReal = real
	rootMu.Unlock()
	return real, nil
}

func GetHomeRoot() string {
	rootMu.RLock()
	defer rootMu.RUnlock()
	return configuredReal
}

func resolveWithinRoot(target string) (string, string, error) {
	rootMu.RLock()
	root := configuredRoot
	rootReal := configuredReal
	rootMu.RUnlock()

	if strings.TrimSpace(target) == "" {
		target = root
	}
	resolved := filepath.Clean(target)
	real, err := filepath.EvalSymlinks(resolved)
	if err != nil {
		return "", "", err
	}
	if real != rootReal && !strings.HasPrefix(real, rootReal+string(filepath.Separator)) {
		return "", "", errors.New("path must be inside the configured base path")
	}
	return real, rootReal, nil
}

func ListFolder(inputPath string) (FolderListing, error) {
	current, rootReal, err := resolveWithinRoot(inputPath)
	if err != nil {
		return FolderListing{}, err
	}

	dirEntries, err := os.ReadDir(current)
	if err != nil {
		return FolderListing{}, err
	}

	entries := make([]FolderEntry, 0, len(dirEntries))
	for _, entry := range dirEntries {
		if strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		kind := "file"
		if entry.IsDir() {
			kind = "directory"
		}
		entries = append(entries, FolderEntry{
			Name: entry.Name(),
			Path: filepath.Join(current, entry.Name()),
			Kind: kind,
		})
	}

	sort.Slice(entries, func(i, j int) bool {
		if entries[i].Kind != entries[j].Kind {
			return entries[i].Kind == "directory"
		}
		return strings.ToLower(entries[i].Name) < strings.ToLower(entries[j].Name)
	})

	var parent *string
	if current != rootReal {
		p := filepath.Dir(current)
		parent = &p
	}

	return FolderListing{
		Root:    rootReal,
		Path:    current,
		Parent:  parent,
		Entries: entries,
	}, nil
}

func FileInfoKind(info fs.FileInfo) string {
	if info.IsDir() {
		return "directory"
	}
	return "file"
}
